README.md
buildc
======

An assistant tool for c project building and packing management

User Guide
-------------
For more information, please open and read the project wiki. cbehave user guide is ready for you.


ChangeLog
# 
# ChangeLog for buildc
#
2014-04-10 Wtz <wtz1989227@gmail.com>
    * add the SVN repository version change processing logic
    * add setting svn user name and password in the .buildc.rc
    * add --user and --password parameter when executing "buildc pack build/source"
    * add default src/lib directory clean up in the packaging process
    * add filtering out .svn path in the packaging process
    * add displaying output information of make command in the packaging process
    * version 0.3.2 released

2013-07-15 Tony Bai <bigwhite.cn@gmail.com>
    * add README.md
    * version 0.3.1 released

2013-07-12 Wtz <wtz1989227@gmail.com>
    * cache_svn_tree.py: fix the problem when display information
    * improve build_tree() to generate valid cache data nodes
    * cache_svn_tree.py: fix the problem that get_cache_libs() returns duplicate data
    * add --config parameter that specifies the name of project related configuration file
    * add var=name parameter that specifies the name of project related configuration file

2013-06-21 Wtz <wtz1989227@gmail.com>
    * svn_local_oper.py: fix the problem when updating svn

2013-05-16 Wtz <wtz1989227@gmail.com>
    * makerules.py: fix the problem when checking archive file

2013-05-10 Wtz <wtz1989227@gmail.com>
    * redefine the semantic meaning of buildc cache init/update/upgrade/remove
    * redefine the semantic meaning of buildc config make
    * refactor the code base using OO paradigm
    * version 0.3.0 released

2013-05-10 Wtz <wtz1989227@gmail.com>
    * check whether "file_to_validate" in .buildc.rc exists when cache upgrade command is executed
    * check whetehr the Make.rules file is empty when executing "config make"
    * add VERSION file which contains buildc version information into the package
    * version 0.2.3 released

2013-01-29 Wtz <wtz1989227@gmail.com>
    * fix buildc __check_buildc_cfg() function, the function to increase the library path existence check
    * fix buildc pack build command

2013-01-08 Wtz <wtz1989227@gmail.com>
    * add pack source command
    * version 0.2.2 released

2012-12-19 Wtz <wtz1989227@gmail.com>
    * add the parameter --ignore-error for buildc cache command

2012-12-05 Wtz <wtz1989227@gmail.com>
    * fix buildc cache manage problem to make these command work according to the cmode
    * version 0.2.1 released

2012-11-28 Wtz <wtz1989227@gmail.com>
    * fix the bug when use buildc pack build --cmode=32-bit

2012-11-01 Wtz <wtz1989227@gmail.com>
    * support packing the dependent dynamic-link library according to the platform which the target would run on
	* version 0.2.0 released

2012-07-25 Wtz <wtz1989227@gmail.com>
    * fix a bug when generated VERSION file

2012-07-18 Wtz <wtz1989227@gmail.com>
    * change the meaning of 'buildc pack build' support for multiple programs to compile and package
    * add packaging process to generate the package information file VERSION
    * supply three detailed example of setup.py.in, deps_check.py.in and env_gen.py.in
    * version 0.1.9 released

2012-07-02 Wtz <wtz1989227@gmail.com>
	* remove FOPTIMIZE from Make.rules.in template
	* add backup command and logging facility in setup.py.in template
	* change install arg from 'full' to 'all' in setup.py.in
    * supply an detailed example of deps_check.py.in
	* version 0.1.8 released

2012-06-04 Wtz <wtz1989227@gmail.com>
    * fix the bug when parsing libname like "libglib-2.0.so"

2012-04-25 Tony Bai <bigwhite.cn@gmail.com>
    * fix the bug when generate make.rules on redhat 64-bit linux

2012-04-18 Tony Bai <bigwhite.cn@gmail.com>
    * change the meaning of 'buildc init'
	* add 'buildc config init' command
	* change 'buildc config-make' to 'buildc config make'
	* version 0.1.7 released

2012-04-18 Tony Bai <bigwhite.cn@gmail.com>
    * add __check_buildc_rc and modify the implementation of cache upgrade
	* adjust the structure of source code
	* move all files in samples to templates and remove samples folder
	* version 0.1.6 released

2012-04-13 Tony Bai <bigwhite.cn@gmail.com>
    * add STATIC_LIBS and DYNAMIC_LIBS
	* fix a bug when config multiple custom libs
	* add SYSTEM_LIBS
	* version 0.1.5 released

2012-04-12 Tony Bai <bigwhite.cn@gmail.com>
    * Version 0.1.4 released
    * fix a bug in command.execute when command executing result is over 256

2012-04-10 Tony Bai <bigwhite.cn@gmail.com>
    * Version 0.1.3 released

2012-03-21 Tony Bai <bigwhite.cn@gmail.com>
    * transpose $(LIBS_DEPEND) and $(CUSTOM_LIBS) in Make.rules.in

2012-02-16 Tony Bai <bigwhite.cn@gmail.com>
    * Version 0.1.2 released
    * add env_gen.py.sample, deps_check.py.sample , layout.cfg.sample and setup.py.sample

2012-02-10 Tony Bai <bigwhite.cn@gmail.com>
    * add pack [create, build, clean, upload] command for setup project

2012-02-08 Tony Bai <bigwhite.cn@gmail.com>
    * Version 0.1.1 released

2012-02-08 Tony Bai <bigwhite.cn@gmail.com>
   * In order to fix the bug that on x86 solaris make tool can not recognize the '~' in path literal, we expand lib cache path when config-make

2012-02-07 Tony Bai <bigwhite.cn@gmail.com>
   * add support for specifying the install path
   * Add setup.sh and ChangeLog
   * Version 0.1.1 released

2011-12-08 Tony Bai <bigwhite.cn@gmail.com>
    * Version 0.1.0 released


core/glo.py
#! /usr/bin/env python
import os
import sys
from utils.errnos import Errors
from utils.util import Util
from utils.system_local_info import SystemLocalInfo

class Glo(object):
    default_includes = ''
    default_libs     = ''

    DOT_BUILDC_RC = Util.usr_home() + os.sep + '.buildc.rc'
    DOT_BUILDC_REPOSITORY = Util.usr_home() + os.sep + '.buildc.repository'
    BUILDC_CFG_PATH = './buildc.cfg'
    MAKE_RULES_PATH = './Make.rules'
    MAKE_RULES_TEMP_PATH = './.Make.rules'

    core_home   = os.path.dirname(os.path.realpath(__file__))
    buildc_home = core_home[:str(core_home).rfind(os.sep)]
    MAKERULES_TPL_PATH = buildc_home + os.sep + 'templates' + os.sep + 'Make.rules.in'
    SETUP_CFG_TPL_PATH = buildc_home + os.sep + 'templates' + os.sep + 'setup.cfg.in'
    SETUP_PY_TPL_PATH = buildc_home + os.sep + 'templates' + os.sep + 'setup.py.in'
    LAYOUT_CFG_TPL_PATH = buildc_home + os.sep + 'templates' + os.sep + 'layout.cfg.in'
    CHECKC_CFG_TPL_PATH = buildc_home + os.sep + 'templates' + os.sep + 'checkc.cfg.in'
    ENV_GEN_PY_TPL_PATH = buildc_home + os.sep + 'templates' + os.sep + 'env_gen.py.in'
    BUILDC_CFG_TPL_PATH = buildc_home + os.sep + 'templates' + os.sep + 'buildc.cfg.in'
    BUILDC_RC_TPL_PATH = buildc_home + os.sep + 'templates' + os.sep + 'buildc.rc.in'

    ONE_TUPLE   = 1
    TWO_TUPLE   = 2
    THREE_TUPLE = 3

    BIT32 = '32-bit'
    BIT64 = '64-bit'
    CPU    = SystemLocalInfo.cpu()
    SYSTEM = SystemLocalInfo.system()
    PACK_SUFFIX = ""

    VAR_STR = ""
    SOURCE_SVN_USER   = None
    SOURCE_SVN_PASSWD = None

    @staticmethod
    def var_str():
        return Glo.VAR_STR

    @staticmethod
    def source_svn_user():
        return Glo.SOURCE_SVN_USER

    @staticmethod
    def source_svn_passwd():
        return Glo.SOURCE_SVN_PASSWD

    @staticmethod
    def dot_buildc_rc_path():
        return Glo.DOT_BUILDC_RC

    @staticmethod
    def dot_buildc_repository_path():
        return Glo.DOT_BUILDC_REPOSITORY

    @staticmethod
    def buildc_cfg_path():
        return Glo.BUILDC_CFG_PATH

    @staticmethod
    def make_rules_path():
        return Glo.MAKE_RULES_PATH

    @staticmethod
    def make_rules_temp_path():
        return Glo.MAKE_RULES_TEMP_PATH

    @staticmethod
    def make_rules_tpl_path():
        return Glo.MAKERULES_TPL_PATH

    @staticmethod
    def setup_cfg_tpl_path():
        return Glo.SETUP_CFG_TPL_PATH

    @staticmethod
    def setup_py_tpl_path():
        return Glo.SETUP_PY_TPL_PATH

    @staticmethod
    def layout_cfg_tpl_path():
        return Glo.LAYOUT_CFG_TPL_PATH

    @staticmethod
    def checkc_cfg_tpl_path():
        return Glo.CHECKC_CFG_TPL_PATH

    @staticmethod
    def env_gen_py_tpl_path():
        return Glo.ENV_GEN_PY_TPL_PATH

    @staticmethod
    def buildc_cfg_tpl_path():
        return Glo.BUILDC_CFG_TPL_PATH

    @staticmethod
    def buildc_rc_tpl_path():
        return Glo.BUILDC_RC_TPL_PATH

    @staticmethod
    def add_backlash(string):
        return string.replace("/", "\/")

    @staticmethod
    def libname2compile_option(libname):
        return ((libname.replace("lib", "", 1)).replace(".a", "")).replace(".so", "")

    @staticmethod
    def is_static_lib(libfilename):
        return libfilename.endswith(".a")

    @staticmethod
    def get_dependent_name_and_version(dependence):
        dep_libname    = None
        dep_libversion = None
        dep_tagfile    = None

        if len(dependence) == Glo.THREE_TUPLE:
            (dep_libname, dep_libversion, dep_tagfile) = dependence
        elif len(dependence) == Glo.TWO_TUPLE:
            (dep_libname, dep_libversion) = dependence
        else:
            print 'dependence args invalid'
            sys.exit(Errors.args_invalid)

        return (dep_libname, dep_libversion, dep_tagfile)

    @staticmethod
    def get_repository(item):
        repository = None
        if len(item) == Glo.ONE_TUPLE:
            if isinstance(item[0], tuple):
                repository = item[0]
            else:
                repository = item
        elif len(item) == Glo.TWO_TUPLE:
            if isinstance(item[0], tuple):
                repository = item[0]
            else:
                repository = item
        else:
            print 'tuple number invalid in .buildc.rc'
            sys.exit(Errors.tuple_number_invalid)

        return repository

    @staticmethod
    def get_local_cache_path(svn_path, repositories):
        url        = None
        cache_path = None

        for item in repositories:
            repository = Glo.get_repository(item)

            if len(repository) == Glo.ONE_TUPLE:
                url = repository[0]
                if svn_path == url:
                    cache_path = '~/buildc_libs/'
                    cache_path += url[str(url).rfind('/')+1:]
                    return os.path.abspath(os.path.expanduser(cache_path))
            elif len(repository) == Glo.TWO_TUPLE or len(repository) == Glo.THREE_TUPLE:
                (url, cache_path) = (repository[0], repository[1])
                if svn_path == url:
                    return os.path.abspath(os.path.expanduser(cache_path))
            else:
                print 'tuple number invalid in .buildc.rc'
                sys.exit(Errors.tuple_number_invalid)

        return None
